src/rpp/rpp/operators/retry_when.hpp

```diff
@@ -0,0 +1,190 @@
+//                  ReactivePlusPlus library
+//
+//          Copyright Aleksey Loginov 2023 - present.
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//          https://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/victimsnino/ReactivePlusPlus
+//
+
+#pragma once
+
+#include <rpp/observables/fwd.hpp>
+#include <rpp/operators/fwd.hpp>
+
+#include <rpp/defs.hpp>
+#include <rpp/operators/details/strategy.hpp>
+#include <rpp/utils/constraints.hpp>
+#include <rpp/utils/utils.hpp>
+
+namespace rpp::operators::details
+{
+    template<rpp::constraint::observer TObserver,
+             typename TObservable,
+             typename TNotifier>
+    struct retry_when_state final : public rpp::composite_disposable
+    {
+        retry_when_state(TObserver&& observer, const TObservable& observable, const TNotifier& notifier)
+            : observer(std::move(observer))
+            , observable(observable)
+            , notifier(notifier)
+        {
+        }
+
+        RPP_NO_UNIQUE_ADDRESS TObserver   observer;
+        RPP_NO_UNIQUE_ADDRESS TObservable observable;
+        RPP_NO_UNIQUE_ADDRESS TNotifier   notifier;
+
+        std::atomic_bool is_inside_drain{};
+    };
+
+    template<rpp::constraint::observer TObserver, typename TObservable, typename TNotifier>
+    void drain(const std::shared_ptr<retry_when_state<TObserver, TObservable, TNotifier>>& state);
+
+    template<rpp::constraint::observer TObserver,
+             typename TObservable,
+             typename TNotifier>
+    struct retry_when_impl_inner_strategy
+    {
+        using preferred_disposable_strategy = rpp::details::observers::none_disposable_strategy;
+
+        std::shared_ptr<retry_when_state<TObserver, TObservable, TNotifier>> state;
+        mutable bool                                                         locally_disposed{};
+
+        template<typename T>
+        void on_next(T&&) const
+        {
+            locally_disposed = true;
+            drain<TObserver, TObservable, TNotifier>(state);
+        }
+
+        void on_error(const std::exception_ptr& err) const
+        {
+            locally_disposed = true;
+            state->observer.on_error(err);
+        }
+
+        void on_completed() const
+        {
+            locally_disposed = true;
+            state->observer.on_completed();
+        }
+
+        void set_upstream(const disposable_wrapper& d) { state->observer.set_upstream(d); }
+
+        bool is_disposed() const { return locally_disposed || state->is_disposed(); }
+    };
```

Comment:
@CorentinBT , it is correct comment. Comparing to on_error in retry and on_completed in repeat, there on_next can be called multiple times, so, you also use locally_disposed inside on_next to avoid it